### ISSUE TITLE
Not correct service map with calico eBPF mode (FELIX_BPFCONNECTTIMELOADBALANCING), linkerd 

### ISSUE THREAD

**Issue by MemberIT** [coroot/coroot#384]
The service map is not generated correctly when using linkerd and calico with eBPF mode enabled as the CNI with settings:
```yaml
        - name: FELIX_BPFCONNECTTIMELOADBALANCING
          value: Disabled
        - name: FELIX_BPFHOSTNETWORKEDNATWITHOUTCTLB
          value: Enabled
```
for smi correct work (https://github.com/projectcalico/calico/issues/8495) - GRPC balancing, dynamyc routing, etc
Similar configurations are available in Cilium for proper SMI (Istio, Linkerd) functionality.
Links on the topic of Cilium + Linkerd, Istio:
[Problem+Resolve](https://linkerd.buoyant.io/t/linkerd-dynamic-request-routing-support-with-cilium-socket-lb/99), [Resolve for Istio](https://docs.cilium.io/en/stable/network/servicemesh/istio/)

It looks very similar to the issue https://github.com/coroot/coroot/issues/3, after which support for Cilium was added.

You can reproduce the issue by deploying coroot in a Kubernetes cluster (CNI: calico eBPF mode), linkerd and checking the service map between coroot and clickhouse.

**Coroot & CH - Not correct service map (service - ebpf):**
![image](https://github.com/user-attachments/assets/13ae0596-81f0-4090-830f-c246fb9d9c59)
**Test apps - correct service map (smi linkerd):**
![image](https://github.com/user-attachments/assets/152659df-6bd5-41ed-b356-e6867f27c732)

**Technical Details:**
Kubernetes: v1.29.10 (k3s)
Calico: v3.27.4 (eBPF mode) settings:
```yaml
        - name: FELIX_BPFCONNECTTIMELOADBALANCING
          value: Disabled
        - name: FELIX_BPFHOSTNETWORKEDNATWITHOUTCTLB
          value: Enabled
```
calicoctl: v3.27.4
linkerd: edge-24.10.5
step: v0.28.0
helm: 3.16.3
coroot: 2.15.13
podinfo

**How to Reproduce:**

*Requirements: VM like Hetzner CX32* 

1. Install k3s, cailco, helm:
```bash
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION='v1.29.10+k3s1' K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="--flannel-backend=none --cluster-cidr=192.168.0.0/16 --disable-network-policy --disable=traefik" sh -
kubectl apply -f https://raw.githubusercontent.com/projectcalico/calico/v3.27.4/manifests/calico.yaml
export KUBECONFIG=/etc/rancher/k3s/k3s.yaml
kubectl completion bash  > /etc/bash_completion.d/kubectl
wget https://get.helm.sh/helm-v3.16.3-linux-amd64.tar.gz
tar -xzvf helm-v3.16.3-linux-amd64.tar.gz
mv linux-amd64/helm /usr/local/bin/
chmod 755 /usr/local/bin/helm
rm -rf linux-amd64/
chown root:root /usr/local/bin/helm
helm completion bash > /etc/bash_completion.d/helm
curl -L https://github.com/projectcalico/calico/releases/download/v3.27.4/calicoctl-linux-amd64 -o calicoctl
chown root:root calicoctl
mv calicoctl /usr/local/bin/
chmod +x /usr/local/bin/calicoctl
cat <<EOF > calico-config.yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: kubernetes-services-endpoint
  namespace: kube-system
data:
  KUBERNETES_SERVICE_HOST: '127.0.0.1'
  KUBERNETES_SERVICE_PORT: '6443'
EOF
kubectl create -f calico-config.yaml
kubectl delete pod -n kube-system -l k8s-app=calico-node
kubectl delete pod -n kube-system -l k8s-app=calico-kube-controllers
sleep 5
calicoctl patch felixconfiguration default --patch='{"spec": {"bpfEnabled": true}}'
```
2. Install step, linkerd:
```bash
wget https://dl.smallstep.com/gh-release/cli/gh-release-header/v0.28.0/step-cli_0.28.0-1_amd64.deb
dpkg -i step-cli_0.28.0-1_amd64.deb
step certificate create root.linkerd.cluster.local ca.crt ca.key --profile root-ca --no-password --insecure
step certificate create identity.linkerd.cluster.local issuer.crt issuer.key --profile intermediate-ca --not-after 8760h --no-password --insecure --ca ca.crt --ca-key ca.key
helm repo add linkerd-edge https://helm.linkerd.io/edge
helm repo update
helm install linkerd-crds linkerd-edge/linkerd-crds --version 2024.10.5 -n linkerd --create-namespace
helm install linkerd-control-plane   --version 2024.10.5 -n linkerd   --set-file identityTrustAnchorsPEM=ca.crt   --set-file identity.issuer.tls.crtPEM=issuer.crt   --set-file identity.issuer.tls.keyPEM=issuer.key   linkerd-edge/linkerd-control-plane
wget https://github.com/linkerd/linkerd2/releases/download/edge-24.10.5/linkerd2-cli-edge-24.10.5-linux-amd64
mv linkerd2-cli-edge-24.10.5-linux-amd64 /usr/local/bin/linkerd
chmod +x /usr/local/bin/linkerd
linkerd completion bash > /etc/bash_completion.d/linkerd
sleep 60
linkerd check
```
3. Install apps for check SMI (podinfo):
```bash
kubectl create ns test --dry-run=client -o yaml   | linkerd inject -   | kubectl apply -f -
helm repo add podinfo https://stefanprodan.github.io/podinfo
helm install backend-a -n test   --set ui.message='A backend' podinfo/podinfo
helm install backend-b -n test   --set ui.message='B backend' podinfo/podinfo
helm install frontend -n test   --set backend=http://backend-a-podinfo:9898/env podinfo/podinfo
```
4. Test dynamic routing - NOT WORKING ([details here](https://linkerd.io/2-edge/tasks/configuring-dynamic-request-routing/))
```
kubectl -n test port-forward svc/frontend-podinfo 9898 &
cat <<EOF | kubectl -n test apply -f -
apiVersion: policy.linkerd.io/v1beta2
kind: HTTPRoute
metadata:
  name: backend-router
  namespace: test
spec:
  parentRefs:
    - name: backend-a-podinfo
      kind: Service
      group: core
      port: 9898
  rules: 
    - matches:  
      - headers:
        - name: "x-request-id"
          value: "alternative"
      backendRefs:
        - name: "backend-b-podinfo"
          port: 9898
    - backendRefs:
      - name: "backend-a-podinfo"
        port: 9898
EOF
```
Test without header - A backend (**Correct**)
```bash
# curl -sX POST localhost:9898/echo   | grep -o 'PODINFO_UI_MESSAGE=. backend'
Handling connection for 9898
PODINFO_UI_MESSAGE=A backend
```
Test with header - A backend (**Not correct**)
```bash
# curl -sX POST   -H 'x-request-id: alternative'   localhost:9898/echo   | grep -o 'PODINFO_UI_MESSAGE=. backend'
Handling connection for 9898
PODINFO_UI_MESSAGE=A backend
```
5. Disable ebpf mode for pods with linkerd

Add ENVs for calico-node daemonset:
```yaml
        - name: FELIX_BPFCONNECTTIMELOADBALANCING
          value: Disabled
        - name: FELIX_BPFHOSTNETWORKEDNATWITHOUTCTLB
          value: Enabled
```
```bash
# kubectl  -n kube-system edit daemonsets.apps calico-node
# kubectl -n kube-system -l k8s-app=calico-node get po 
NAME                READY   STATUS    RESTARTS   AGE
calico-node-rrv97   1/1     Running   0          29s
```
6. Test dynamic routing - WORKING  ([details here](https://linkerd.io/2-edge/tasks/configuring-dynamic-request-routing/))

Test without header - A backend (**Correct**)
```bash
# curl -sX POST localhost:9898/echo   | grep -o 'PODINFO_UI_MESSAGE=. backend'
Handling connection for 9898
PODINFO_UI_MESSAGE=A backend
```
Test with header - B backend (**Correct**)
```bash
# curl -sX POST   -H 'x-request-id: alternative'   localhost:9898/echo   | grep -o 'PODINFO_UI_MESSAGE=. backend'
Handling connection for 9898
PODINFO_UI_MESSAGE=B backend
```
**linkerd is working**

7. And now you can install coroot for reproduce problem:

Install coroot:
```bash
helm repo add coroot https://coroot.github.io/helm-charts
helm repo update coroot 
helm install --namespace coroot --create-namespace coroot coroot/coroot
```
Wait few minutes and forward port for web coroot:
```
kubectl --namespace coroot port-forward service/coroot 8080:8080 --address 0.0.0.0 &
```
Open link http://IP_VM:8080/ and check the service map for monitoring

**Comment by MemberIT** [comment]
Fixed in version 1.6.x

**Comment by MemberIT** [comment]
Fixed in version 1.6.x
